sparse_strips/vello_common/src/recording.rs

```diff
@@ -0,0 +1,410 @@
+    /// ```ignore
+    /// let mut recording = scene.record(|ctx| {
+    ///     ctx.fill_rect(&Rect::new(0.0, 0.0, 100.0, 100.0));
+    /// });
+    ///
+    /// // Generate strips explicitly
+    /// scene.prepare_recording(&mut recording);
+    ///
+    /// // Now render using pre-generated strips
+    /// scene.render_recording(&mut recording);
+    /// ```
+    fn prepare_recording(&mut self, recording: &mut Recording) {
+        if !recording.has_cached_strips() {
+            let (strips, alphas, strip_ranges) =
+                self.generate_strips_from_commands(&recording.commands);
+            recording.set_cached_strips(strips.into_boxed_slice(), alphas.into_boxed_slice());
+            recording.set_strip_ranges(strip_ranges);
+        }
+    }
+
+    /// Render using a recording (caches strips on first use)
+    ///
+    /// This method executes a previously recorded sequence of operations.
+    /// On first use, it will generate and cache the necessary rendering data.
+    /// Subsequent calls will reuse the cached data for better performance.
+    fn render_recording(&mut self, recording: &mut Recording) {
+        self.prepare_recording(recording);
+        self.execute_recording(recording);
+    }
+
+    /// Record and prepare strips immediately
+    ///
+    /// This is a convenience method that combines `record` and `prepare_recording`.
+    /// Use this when you want to pre-generate strips without rendering yet.
+    ///
+    /// # Example
+    /// ```ignore
+    /// let recording = scene.record_and_prepare(|ctx| {
+    ///     ctx.fill_rect(&Rect::new(0.0, 0.0, 100.0, 100.0));
+    /// });
+    /// // Strips are now generated and cached
+    /// ```
+    fn record_and_prepare<F>(&mut self, f: F) -> Recording
+    where
+        F: FnOnce(&mut Recorder<'_>),
+    {
+        let mut recording = self.record(f);
+        self.prepare_recording(&mut recording);
+        recording
+    }
+
+    /// Record and render immediately
+    ///
+    /// This is a convenience method that combines `record` and `render_recording`.
+    /// Use this when you want to execute commands immediately but also keep
+    /// the recording for potential reuse.
+    fn record_and_render<F>(&mut self, f: F) -> Recording
+    where
+        F: FnOnce(&mut Recorder<'_>),
+    {
+        let mut recording = self.record(f);
+        self.render_recording(&mut recording);
+        recording
+    }
+
+    /// Render recording if available, otherwise record and render
+    ///
+    /// This method provides a convenient way to handle optional recordings.
+    /// If a recording is provided, it will be rendered. Otherwise, it will
+    /// record and render the provided commands.
+    fn render_or_record<F>(&mut self, recording: Option<&mut Recording>, f: F) -> Recording
+    where
+        F: FnOnce(&mut Recorder<'_>),
+    {
+        if let Some(existing) = recording {
+            self.render_recording(existing);
+            existing.clone()
```

Comment:
I think this line is unexpected from the consumers POV. This could be cloning 3 million alphas and cause significant performance implications for the consumer. I'm wondering whether we could adjust the signature to take an `Option<Recording>` so this isn't required or remove it completely (can consumers emulate this function themselves?)